libraries/standard/http/src/http_client.c

```diff
@@ -245,6 +245,986 @@ static int findHeaderValueParserCallback( http_parser * pHttpParser,
+ *
+ * The follow is an example of a Transfer-Encoding chunked response:
+ *
+ * HTTP/1.1 200 OK\r\n
+ * Content-Type: text/plain\r\n
+ * Transfer-Encoding: chunked\r\n
+ * \r\n
+ * d\r\n
+ * Hello World! \r\n
+ * 7\r\n
+ * I am a \r\n
+ * a\r\n
+ * developer.\r\n
+ * 0\r\n
+ * \r\n
+ *
+ * The first invocation of this callback will contain @p pLoc = "Hello World!"
+ * and @p length = 13.
+ * The second invocation of this callback will contain @p pLoc = "I am a " and
+ * @p length = 7.
+ * The third invocation of this callback will contain @p pLoc = "developer." and
+ * @p length = 10.
+ *
+ * @param[in] pHttpParser Parsing object containing state and callback context.
+ * @param[in] pLoc - Pointer to the body string in the response message buffer.
+ * @param[in] length - The length of the body found.
+ *
+ * @return Zero to continue parsing. All other return values will stop parsing
+ * and http_parser_execute() will return with status HPE_CB_body.
+ */
+static int httpParserOnBodyCallback( http_parser * pHttpParser,
+                                     const char * pLoc,
+                                     size_t length );
+
+/**
+ * @brief Callback invoked during http_parser_execute() to indicate the the
+ * completion of an HTTP response message.
+ *
+ * When there is no response body, the end of the response message is when the
+ * headers end. This is indicated by another "\r\n" after the final header line.
+ *
+ * When there is response body, the end of the response message is when the
+ * full "Content-Length" value is parsed following the end of the headers. If
+ * there is no Content-Length header, then http_parser_execute() expects a
+ * zero length-ed parsing data to indicate the end of the response.
+ *
+ * For a "Transfer-Encoding: chunked" type of response message, the complete
+ * response message is signalled by a terminating chunk header with length zero.
+ *
+ * See https://github.com/nodejs/http-parser for more information.
+ *
+ * @param[in] pHttpParser Parsing object containing state and callback context.
+ *
+ * @return Zero to continue parsing. All other return values will stop parsing
+ * and http_parser_execute() will return with status HPE_CB_message_complete.
+ */
+static int httpParserOnMessageCompleteCallback( http_parser * pHttpParser );
+
+/**
+ * @brief When a complete header is found the HTTP response header count
+ * increases and the application is notified.
+ *
+ * This function is invoked only in callbacks that could follow
+ * #httpParserOnHeaderValueCallback. These callbacks are
+ * #httpParserOnHeaderFieldCallback and #httpParserOnHeadersCompleteCallback.
+ *
+ * Because of #httpParserOnHeadersCompleteCallback and
+ * #httpParserOnHeaderValueCallback having the possibility of being invoked
+ * multiple times for partial header fields and values, it is not known if a
+ * header field and value pair is complete until
+ * httpParserOnHeaderValueCallback() does not get called in succession.
```

Comment:
Does this sentence point to the reason for this API to be invoked by callbacks that follow #httpParserOnHeaderValueCallback? I didn't completely understand this.

Reply:
yes, let me reword a bit.